### PR TITLE
Fixing errors display when using options '--ns' or '--ds'

### LINKED ISSUE #358: Incorrect header placement when using options '--ns' or '--ds'
Currently, in some cases the messages header is displayed too late (i.e. after some calls to Engine are done). Specifically, when using either option `--ns` or `--ds,` it can be seen in case of any system error (`SYSTEM:UNSPECIFIED`), e.g:

```
$ zonemaster-cli --show-testcase --no-ipv6 --test basic/basic01 zonemaster.net --ns 'ns1.zonemaster.net'
   0.72 ERROR    UNSPECIFIED    The fake delegation of domain zonemaster.net includes an in-zone name server ns1.zonemaster.net without mandatory glue (without IP address).
Seconds Level    Testcase       Message
======= ======== ============== =======
```

I suggest that the [code displaying the header](https://github.com/zonemaster/zonemaster-cli/blob/v6.0.3/lib/Zonemaster/CLI.pm#L587-L619) is moved earlier in the code. A more appropriate place would be just after the successful normalization of the input domain name (e.g. [line 565](https://github.com/zonemaster/zonemaster-cli/blob/v6.0.3/lib/Zonemaster/CLI.pm#L565) or [line 578](https://github.com/zonemaster/zonemaster-cli/blob/v6.0.3/lib/Zonemaster/CLI.pm#L578)).

### PR DESCRIPTION
**Pull request by MichaelTimbert**
## Purpose

This separate CLI errors from Engine errors.
Errors on CLI argument are catch before displaying the header.

## Context

Fixes #358 

## Changes

I have split CLI argument verification from Engine settings.

I have also add regex validation for '--ds' arguments

## How to test this PR

Stress test zonemaster-cli with various '--ns' and '--ds' arguments.
Errors from command line input validation must be display **before** any header.
Errors from the Engine must be formatted correctly **after** the header.

here some example to cmd to run:
```
zonemaster-cli --show-testcase --no-ipv6 --test basic/basic01 zonemaster.net --ns 'ns1.zonemaster.net'
zonemaster-cli --show-testcase  example.fr --ns 'ns1.zonemaster..fr.'
zonemaster-cli --show-testcase  example.fr --ns 'ns1.zonemaster.fr/542.0.0.1'
zonemaster-cli --show-testcase  example.fr --ns '/215.4.6.1'
zonemaster-cli --show-testcase afnic.fr --ds "32674 13 2 12CCEEFE3ECD007C"
zonemaster-cli --show-testcase afnic.fr --ds "32674,13,2,NOTHEXADECIMAL"
```

### CODE REVIEW

File: lib/Zonemaster/CLI.pm [hunk 1]
```diff
@@ -839,16 +851,49 @@ sub add_fake_delegation {
...
+
+}
+
+sub check_fake_ds {
+    my ( $self, $domain ) = @_;
+    my @data;
+
+    foreach my $str ( @{ $self->ds } ) {
+
+        if ( not $str =~ /$DS_RE/g ) {
```
Comment (marc-vanderwal) [inline lib/Zonemaster/CLI.pm:876]: Is the `g` flag really necessary for the regular expression?

**Comment by matsduf** [comment]
This PR is based on develop branch before the last release. I suggest that it is rebased on latest master or develop branch.

**Review by matsduf — COMMENTED** [review]
The changes are fine. When the other other review comments are handled I can approve.

File: lib/Zonemaster/CLI.pm [hunk 2]
```diff
@@ -839,16 +847,46 @@ sub add_fake_delegation {
...
+                ( $ip =~ /($IPV4_RE)/ && Net::IP::XS::ip_is_ipv4( $ip ) )
                 or
-                 ( $ip =~ /($IPV6_RE)/ && Net::IP::XS::ip_is_ipv6( $ip ) )
-            ) {
-                push @{ $data{ $name } }, $ip;
-            }
-            else {
+                ( $ip =~ /($IPV6_RE)/ && Net::IP::XS::ip_is_ipv6( $ip ) )
+            )
+	    {
```
Comment (matsduf) [inline lib/Zonemaster/CLI.pm:855]: I suggest to call Zonemaster::Engine::Validation::validate_ipv4 and Zonemaster::Engine::Validation::validate_ipv6, respectively, instead.

https://github.com/zonemaster/zonemaster-engine/blob/master/lib/Zonemaster/Engine/Validation.pm#L28-L57

**Review by matsduf — DISMISSED** [review]

File: lib/Zonemaster/CLI.pm [hunk 3]
```diff
@@ -639,12 +640,13 @@ sub run {
...
+
     if ( $self->ns and @{ $self->ns } > 0 ) {
-        $self->add_fake_delegation( $domain );
+        $self->check_fake_delegation( $domain );
     }
 
     if ( $self->ds and @{ $self->ds } ) {
-        $self->add_fake_ds( $domain );
+        $self->check_fake_ds( $domain );
     }
```
[the remark this answers is not in the document]
Comment (MichaelTimbert) [reply lib/Zonemaster/CLI.pm]: I move this input validation code earlier, at line 622, after $domain declaration.

File: lib/Zonemaster/CLI.pm [hunk 4]
```diff
@@ -839,16 +851,49 @@ sub add_fake_delegation {
...
         if ( $ip ) {
             my $net_ip = Net::IP::XS->new( $ip );
-            if ( ( $ip =~ /($IPV4_RE)/ && Net::IP::XS::ip_is_ipv4( $ip ) )
-                or
-                 ( $ip =~ /($IPV6_RE)/ && Net::IP::XS::ip_is_ipv6( $ip ) )
-            ) {
-                push @{ $data{ $name } }, $ip;
-            }
-            else {
+            if ( not(
```
[the remark this answers is not in the document]
Comment (MichaelTimbert) [reply lib/Zonemaster/CLI.pm]: Done

File: lib/Zonemaster/CLI.pm (hunk 1 above)
Comment (MichaelTimbert) [reply lib/Zonemaster/CLI.pm:868]: no, fixed and i am using `unless` instead of `if ( not(`.

File: lib/Zonemaster/CLI.pm [hunk 5]
```diff
@@ -839,16 +847,46 @@ sub add_fake_delegation {
...
+
+    return;
+}
+
+sub check_fake_ds {
+    my ( $self ) = @_;
+
+    foreach my $str ( @{ $self->ds } ) {
+        unless ( $str =~ /$DS_RE/ ) {
+            say STDERR __( "--ds ds data must be in the form \"keytag,algorithm,type,digest\"");
```
Comment (matsduf) [inline lib/Zonemaster/CLI.pm:864]: I suggest
```
say STDERR __( "--ds ds data must be in the form \"keytag,algorithm,type,digest\". E.g. space is not permitted anywhere in the string.");
```
Motivation:

When `dig` lists a DS record is seems to always split the digest in two with a space character. I think that bind accepts the space character. E.g. (see right end):

```
;; ANSWER SECTION:
malmo.se.		3600	IN	DS	24604 8 2 2CDEBE0EAAA9906D8C2025D348C0A984825A2B58F5C67CCB7D6A5429 2C697107
malmo.se.		3600	IN	DS	24604 8 4 4249EE3A52E94F7B42C593D6ABD506A1413B2F06522F0455E58E99B6 263330A224F89576AE56B420603A9BE22EEC42FB
```

**Review by matsduf — DISMISSED** [review]

**Comment by matsduf** [comment]
@MichaelTimbert, can you please handle the conflicts?

**Review by marc-vanderwal — APPROVED** [review]

**Comment by matsduf** [comment]
```
# zonemaster-cli --show-testcase --no-ipv6 --test Basic zonemaster.net --ns ns1.zonemaster.net --level info
   1.11 ERROR    Unspecified    The fake delegation of domain zonemaster.net includes an in-zone name server ns1.zonemaster.net without mandatory glue (without IP address).
  \ 
Seconds Level    Testcase       Message
======= ======== ============== =======
   0.00 INFO     Unspecified    Using version v6.0.0 of the Zonemaster engine.
   0.00 INFO     Basic01        The zone "zonemaster.net" is found.
   0.00 INFO     Basic01        This is a test of an undelegated domain so finding the parent zone is disregarded.
   0.16 CRITICAL Basic02        There is no working name server for "zonemaster.net" so it is unreachable.
   0.16 ERROR    Basic02        Name server "ns1.zonemaster.net" cannot be resolved into an IP address.
```
Why is the first message before the header?

**Review by matsduf — CHANGES_REQUESTED** [review]
```
# zonemaster-cli --show-testcase --no-ipv6 --test Basic zonemaster.net --ns ns1.zonemaster.net --level info
   1.11 ERROR    Unspecified    The fake delegation of domain zonemaster.net includes an in-zone name server ns1.zonemaster.net without mandatory glue (without IP address).
  \ 
Seconds Level    Testcase       Message
======= ======== ============== =======
   0.00 INFO     Unspecified    Using version v6.0.0 of the Zonemaster engine.
   0.00 INFO     Basic01        The zone "zonemaster.net" is found.
   0.00 INFO     Basic01        This is a test of an undelegated domain so finding the parent zone is disregarded.
   0.16 CRITICAL Basic02        There is no working name server for "zonemaster.net" so it is unreachable.
   0.16 ERROR    Basic02        Name server "ns1.zonemaster.net" cannot be resolved into an IP address.
```
Why is the first message before the header?

**Comment by MichaelTimbert** [comment]
> ```
> # zonemaster-cli --show-testcase --no-ipv6 --test Basic zonemaster.net --ns ns1.zonemaster.net --level info
>    1.11 ERROR    Unspecified    The fake delegation of domain zonemaster.net includes an in-zone name server ns1.zonemaster.net without mandatory glue (without IP address).
>   \ 
> Seconds Level    Testcase       Message
> ======= ======== ============== =======
>    0.00 INFO     Unspecified    Using version v6.0.0 of the Zonemaster engine.
>    0.00 INFO     Basic01        The zone "zonemaster.net" is found.
>    0.00 INFO     Basic01        This is a test of an undelegated domain so finding the parent zone is disregarded.
>    0.16 CRITICAL Basic02        There is no working name server for "zonemaster.net" so it is unreachable.
>    0.16 ERROR    Basic02        Name server "ns1.zonemaster.net" cannot be resolved into an IP address.
> ```
> 
> Why is the first message before the header?

I mistakenly moved the `add_fake_*` before displaying the header. it's now fixed.

**Review by matsduf — APPROVED** [review]

**Review by marc-vanderwal — APPROVED** [review]

**Comment by marc-vanderwal** [comment]
Release testing for 2024.2:

When trying the first command, the output doesn’t report any errors, although it should:
```
$ zonemaster-cli --show-testcase --no-ipv6 --test basic/basic01 zonemaster.net --ns 'ns1.zonemaster.net'
  \
Seconds Level    Testcase       Message
======= ======== ============== =======
```

All other command lines listed in the testing procedures show an error, without showing the “seconds, level, testcase, message” header at all, so I assume this works as expected.

I’ve created issue #402 to track this problem.

**Pull request merged**